django/thunderstore/storage/models/blob.py

```diff
@@ -0,0 +1,135 @@
+from hashlib import sha256
+from typing import Optional
+
+from django.conf import settings
+from django.core.files.base import ContentFile
+from django.core.files.storage import get_storage_class
+from django.db import models
+
+from thunderstore.core.mixins import SafeDeleteMixin, TimestampMixin
+
+
+def get_object_file_path(_, filename: str) -> str:
+    return f"blob-storage/sha256/{filename}.sha256.blob"
+
+
+class DataBlob(SafeDeleteMixin):
+    """
+    The DataBlob class is responsible for storing arbitrary blobs of data with
+    automatic deduplication by blob sha256 checksum. It is not interested in
+    any metadata which is not inferable without parsing the data. Additional
+    meta is left for DataBlobReference, which is also interested in contextual
+    information.
+    """
+
+    data = models.FileField(
+        upload_to=get_object_file_path,
+        storage=get_storage_class(settings.BLOB_FILE_STORAGE)(),
+        editable=False,
+        blank=True,
+        null=True,
+    )
+    data_size = models.BigIntegerField()
+    checksum_sha256 = models.CharField(
+        max_length=512,
+        editable=False,
+        null=False,
+        unique=True,
+        db_index=True,
+    )
+
+    @property
+    def data_url(self) -> str:
```

Comment:
Does this actually return a `str` if data is null?

Reply:
I think it throws in that case, but didn't test. It "should" be fine since a situation where the data is null is already an integrity error of some kind (i.e. the deletion was in progress but failed while the db object still remains)